### DISCUSSION TITLE
Dynamic input of topics

### DISCUSSION THREAD

**Discussion by gabrielvictorafreitas** [dotnet/MQTTnet#2015] [General]
Does anyone know if i can have a Dynamic number of topics? My ideia is to configure the topics at appsettings and do like pipe sepparately pattern?

I know that we need to configure something like this:
`var mqttSubscribeOptions = mqttFactory.CreateSubscribeOptionsBuilder() .WithTopicFilter( f => { f.WithTopic("mqttnet/samples/topic/1"); }) .WithTopicFilter( f => { f.WithTopic("mqttnet/samples/topic/2").WithNoLocal(); }) .WithTopicFilter( f => { f.WithTopic("mqttnet/samples/topic/3").WithRetainHandling(MqttRetainHandling.SendAtSubscribe); }) .Build();`

my ideia is to do something like this:

`List topics = new List(topic.Split('|'));

            var mqttSubscribeOptions = factory.CreateSubscribeOptionsBuilder()
                           .WithTopicFilter(
                               f =>
                               {
                                   foreach (var t in topics)
                                   {
                                    
                                   f.WithTopic(t);
                                   }
                               })
                           .Build();`
I know that this configuration is wrong

**Comment by gabrielvictorafreitas** [comment]
Done!
i've created one function to create the topic subscription sepparately:

```csharp
var mqttSubscribeOptions = GenerateSubscriptions();

// Subscribe to all topics
foreach (var topic in mqttSubscribeOptions)
{
    await mqttClient.SubscribeAsync(topic, cancellationToken).ConfigureAwait(false);
}

private List<MqttClientSubscribeOptions> GenerateSubscriptions()
{
    List<MqttClientSubscribeOptions> options = new List<MqttClientSubscribeOptions>();
    List<string> topics = new List<string>(topic.Split('|'));
    var factory = new MqttFactory();

    foreach (var t in topics)
    {
        options.Add(factory.CreateSubscribeOptionsBuilder()
            .WithTopicFilter(
                f =>
                {
                    f.WithTopic(t);
                })
            .Build());

    }

    return options;
}
```

**Reply by SeppPenner** [reply]
Yes, that should work.

**Comment by SeppPenner** [comment]
I guess, you need to do it like that:

```csharp
var subscribeResults = new MqttClientSubscribeResult[topics.Length];

for (var i = 0; i < topics.Length; i++)
{
    var idx = i;
    var mqttClientSubscribeOptions = this.mqttFactory.CreateSubscribeOptionsBuilder()
        .WithTopicFilter(
            f =>
            {
                f.WithTopic(topics[idx]);   
            })
         .Build();
            
    var subscribeResult = await this.mqttClient.SubscribeAsync(mqttClientSubscribeOptions, cancellationToken ?? CancellationToken.None);
    subscribeResults[idx] = subscribeResult;
}
        
return subscribeResults;
```

`topics` is a `List<string>` or `string[]`.

**Reply by gabrielvictorafreitas** [reply]
Yeah! I’ve done something like that and worked perfectly! Thanks for the comment!
